### PR TITLE
Micro-optimize GL temp buffer generation

### PR DESCRIPTION
**Pull request by juj**
Micro-optimize GL temp buffer generation to only be present in code when used.

### CODE REVIEW

File: src/library_webgl.js [hunk 1]
```diff
@@ -8,9 +8,23 @@ var LibraryGL = {
 
   // For functions such as glDrawBuffers, glInvalidateFramebuffer and glInvalidateSubFramebuffer that need to pass a short array to the WebGL API,
   // create a set of short fixed-length arrays to avoid having to generate any garbage when calling those functions.
-  _tempFixedLengthArray__postset: 'for (var i = 0; i < 32; i++) __tempFixedLengthArray.push(new Array(i));',
+  _tempFixedLengthArray__postset: 'for (var ____n = 0; ____n < 32; ++____n) __tempFixedLengthArray.push(new Array(____n));',
```
Comment (kripken) [inline src/library_webgl.js:11]: why this change?

File: src/library_webgl.js [hunk 2]
```diff
@@ -8,9 +8,23 @@ var LibraryGL = {
...
   _tempFixedLengthArray: [],
 
+  _miniTempFloatBuffers: [],
+  _miniTempFloatBuffers__postset:
+      'var __floatBuf = new Float32Array(' + {{{ GL_POOL_TEMP_BUFFERS_SIZE }}} + ');\n'
```
Comment (kripken) [inline src/library_webgl.js:16]: how about `glFloatBuf`?

I don't think we need `__` prefixing for things like this, we don't do that for other things. We just use natural JS names.

And I think mentioning in the name that it is for GL would be clearer.

File: src/library_webgl.js [hunk 3]
```diff
@@ -2223,12 +2216,19 @@ var LibraryGL = {
...
+#endif
   glUniform1iv: function(location, count, value) {
 #if GL_ASSERTIONS
     GL.validateGLObjectID(GL.uniforms, location, 'glUniform1iv', 'location');
     assert((value & 3) == 0, 'Pointer to integer data passed to glUniform1iv must be aligned to four bytes!');
 #endif
 
+#if MIN_WEBGL_VERSION == 2
+    GLctx.uniform1iv(GL.uniforms[location], HEAP32, value>>2, count);
+#else
```
Comment (kripken) [inline src/library_webgl.js:2230]: this part of the change seems unrelated to the temp buffers, or am I missing something?

File: src/settings.js [hunk 4]
```diff
@@ -519,6 +519,10 @@ var GL_SUPPORT_EXPLICIT_SWAP_CONTROL = 0;
 // you can turn it off.
 var GL_POOL_TEMP_BUFFERS = 1;
 
+// Specifies the size of the GL temp buffer pool, in bytes. Must be a multiple
+// of 9 and 16.
+var GL_POOL_TEMP_BUFFERS_SIZE = 2*9*16; // == 288
```
Comment (kripken) [inline src/settings.js:524]: does this need to be a public setting?

looks like this also changes the value, which was 256 - was that intentional?

File: src/library_webgl.js (hunk 2 above)
Comment (kripken) [reply src/library_webgl.js:16]: probably not `glFloatBuf` though as it sounds too much like a real thing. Maybe `tempGLFloatBuf`?

File: src/library_webgl.js (hunk 1 above)
Comment (juj) [reply src/library_webgl.js:11]: The `var` inside the `for` init block applies to the outer (global) scope. Closure does not let the following function compile:
```js
for(var i = 0; i < 10; ++i);
for(var i = 0; i < 10; ++i);
```
but it complains `i` is defined twice.

File: src/library_webgl.js (hunk 2 above)
Comment (juj) [reply src/library_webgl.js:16]: If a non-underscored name is used, that might conflict with a real name that user brings from external code?

I think we used to apply underscores for internal things that are not exposed to users to access?

File: src/library_webgl.js (hunk 3 above)
Comment (juj) [reply src/library_webgl.js:2230]: Yeah, that is to optimize code size on WebGL 2 only code.

File: src/settings.js (hunk 4 above)
Comment (juj) [reply src/settings.js:524]: Changing the value is intentional, note that the check changed from `if (x*9 <= GL_POOL_TEMP_BUFFERS_SIZE)` to `if (x <= GL_POOL_TEMP_BUFFERS_SIZE/9)` to save code size. That requires a value that is multiple of 9 and 16.

It does not need to be a public setting, but I don't think we have a good way for creating file-specific "#defines"?

File: src/library_webgl.js (hunk 1 above)
Comment (kripken) [reply src/library_webgl.js:11]: Hmm that is unfortunate. But is `___n` better? seems like we'll need a new id in each place we have a loop. And multiple underscores specifically worry me as maybe a typo with the number of underscores could happen. How about just `j`?

A singleton `var i;` on top and then loops not having a `var` at all would be more compact, and avoid this. But it's also an error-prone pattern.

File: src/library_webgl.js (hunk 2 above)
Comment (kripken) [reply src/library_webgl.js:16]: I don't think we've had a consistent policy on prefixing of our internal things. Maybe that's in part to it not always being clear if something is internal or not.

But if it's clear this is not for users, then `__tempGLFloatBuf` sounds ok to me.

In theory the cleaner thing for this is to use a closure, but that would increase code size sadly.

File: src/settings.js (hunk 4 above)
Comment (kripken) [reply src/settings.js:524]: sg about the value.

About the setting, this could at least be an internal setting and not a public one. But we can actually set defines in files, {{{ X = Y }}} at the top of the file will just assign that value in the compiler's global scope.

File: src/library_webgl.js (hunk 1 above)
Comment (juj) [reply src/library_webgl.js]: Reverted back to `i` and applied a Closure suppression instead.

File: src/library_webgl.js (hunk 2 above)
Comment (juj) [reply src/library_webgl.js]: Renamed to `__temp_webgl_float_buf` and `__temp_webgl_int_buf`.

File: src/settings.js (hunk 4 above)
Comment (juj) [reply src/settings.js]: That's neat - I think I tried something like that, but botched up the syntax. Moved GL_POOL_TEMP_BUFFERS_SIZE to library_webgl.js now.

File: src/library_webgl.js [hunk 5]
```diff
@@ -4,13 +4,31 @@
...
 
   // For functions such as glDrawBuffers, glInvalidateFramebuffer and glInvalidateSubFramebuffer that need to pass a short array to the WebGL API,
   // create a set of short fixed-length arrays to avoid having to generate any garbage when calling those functions.
-  _tempFixedLengthArray__postset: 'for (var i = 0; i < 32; i++) __tempFixedLengthArray.push(new Array(i));',
+  _tempFixedLengthArray__postset: 'for (var i = 0; i < 32; ++i) __tempFixedLengthArray.push(new Array(i));',
   _tempFixedLengthArray: [],
 
+  _miniTempFloatBuffers: [],
+  _miniTempFloatBuffers__postset:
+      'var __temp_webgl_float_buf = new Float32Array(' + {{{ GL_POOL_TEMP_BUFFERS_SIZE }}} + ');\n'
```
Comment (kripken) [inline src/library_webgl.js:20]: this name uses `x_y` instead of `xY` notation?

also i think this name could be the same as `_miniTempFloatBuffers` but only changed a little, so the connection is clear. like maybe `_miniTempFloatBuffersStorage`?

File: src/library_webgl.js [hunk 6]
```diff
@@ -4,13 +4,31 @@
...
 var LibraryGL = {
 
   // For functions such as glDrawBuffers, glInvalidateFramebuffer and glInvalidateSubFramebuffer that need to pass a short array to the WebGL API,
   // create a set of short fixed-length arrays to avoid having to generate any garbage when calling those functions.
-  _tempFixedLengthArray__postset: 'for (var i = 0; i < 32; i++) __tempFixedLengthArray.push(new Array(i));',
+  _tempFixedLengthArray__postset: 'for (var i = 0; i < 32; ++i) __tempFixedLengthArray.push(new Array(i));',
   _tempFixedLengthArray: [],
 
+  _miniTempFloatBuffers: [],
+  _miniTempFloatBuffers__postset:
```
Comment (kripken) [inline src/library_webgl.js:19]: I still think it would be good to have `GL` in the name of these. it helps when reading the generated code to know what they are immediately, since they are in the global scope with a lot of other stuff.

File: src/library_webgl.js (hunk 5 above)
Comment (juj) [reply src/library_webgl.js:20]: sure

File: src/library_webgl.js (hunk 6 above)
Comment (juj) [reply src/library_webgl.js:19]: check
